criu/memfd-secret.c

```diff
@@ -0,0 +1,282 @@
+{
+	struct memfd_secret_dump_inode *inode;
+
+	list_for_each_entry(inode, &memfd_secret_inodes, list)
+		if ((inode->dev == st->st_dev) && (inode->ino == st->st_ino))
+			return inode;
+
+	inode = xmalloc(sizeof(*inode));
+	if (inode == NULL)
+		return NULL;
+
+	inode->dev = st->st_dev;
+	inode->ino = st->st_ino;
+	inode->id = memfd_secret_inode_ids++;
+
+	if (dump_memfd_secret_inode(lfd, inode, st)) {
+		xfree(inode);
+		return NULL;
+	}
+
+	list_add_tail(&inode->list, &memfd_secret_inodes);
+
+	return inode;
+}
+
+static int dump_one_memfd_secret(int lfd, u32 id, const struct fd_parms *p)
+{
+	MemfdSecretFileEntry msfe = MEMFD_SECRET_FILE_ENTRY__INIT;
+	FileEntry fe = FILE_ENTRY__INIT;
+	struct memfd_secret_dump_inode *inode;
+	struct fd_link _link, *link;
+
+	if (!p->link) {
+		if (fill_fdlink(lfd, p, &_link))
+			return -1;
+		link = &_link;
+	} else
+		link = p->link;
+
+	link_strip_deleted(link); /* link->name: ./secretmem */
+
+	inode = dump_unique_memfd_secret_inode(lfd, &p->stat);
+	if (!inode)
+		return -1;
+
+	msfe.id = id;
+	msfe.flags = p->flags;
+	msfe.pos = p->pos;
+	msfe.fown = (FownEntry *)&p->fown;
+	msfe.inode_id = inode->id;
+
+	fe.type = FD_TYPES__MEMFD_SECRET;
+	fe.id = msfe.id;
+	fe.memfd_secret = &msfe;
+
+	return pb_write_one(img_from_set(glob_imgset, CR_FD_FILES), &fe, PB_FILE);
+}
+
+const struct fdtype_ops memfd_secret_dump_ops = {
+	.type = FD_TYPES__MEMFD_SECRET,
+	.dump = dump_one_memfd_secret,
+};
+
+/* restore */
+
+struct memfd_secret_info {
+	MemfdSecretFileEntry *msfe;
+	struct file_desc d;
+	struct memfd_secret_restore_inode *inode;
+};
+
+static struct memfd_secret_restore_inode *memfd_secret_alloc_inode(int id)
+{
+	struct memfd_secret_restore_inode *inode;
+
+	list_for_each_entry(inode, &memfd_secret_inodes, list)
+		if (inode->id == id)
+			return inode;
+
+	inode = shmalloc(sizeof(*inode));
+	if (!inode)
+		return NULL;
+
+	inode->id = id;
+	mutex_init(&inode->lock);
+	inode->fdstore_id = -1;
+
+	list_add_tail(&inode->list, &memfd_secret_inodes);
+	return inode;
+}
+
+static int memfd_secret_open_inode_nocache(struct memfd_secret_restore_inode *inode)
+{
+#if defined(CONFIG_HAS_MEMFD_SECRET)
+	MemfdSecretInodeEntry *msie = NULL;
+	struct cr_img *img = NULL;
+	int fd = -1;
+	int ret = -1;
+	int flags = 0;
+
+	img = open_image(CR_FD_MEMFD_SECRET_INODE, O_RSTR, inode->id);
+	if (!img)
+		goto out;
+
+	if (pb_read_one(img, &msie, PB_MEMFD_SECRET_INODE) < 0)
+		goto out;
+
+	fd = memfd_secret(flags);
+	if (fd < 0) {
+		pr_perror("Can't create memfd_secret");
+		goto out;
+	}
+
+	if (restore_secretmem_content(fd, msie->secretmem_id, msie->size))
+		goto out;
+
+	if (cr_fchown(fd, msie->uid, msie->gid)) {
+		pr_perror("Can't change uid %d gid %d of memfd-secret", (int)msie->uid, (int)msie->gid);
+		goto out;
+	}
+
+	inode->fdstore_id = fdstore_add(fd);
+	if (inode->fdstore_id < 0)
+		goto out;
+
+	ret = fd;
+	fd = -1;
+
+out:
```

Comment:
Consider using kernel style error handling: https://docs.kernel.org/process/coding-style.html#centralized-exiting-of-functions